### PR TITLE
fold CodeRabbit findings on #2730 (tsk-x6fzgf): Audit lane: add an adversarial-verify stage (refute each finding vs code) before findings become cards [tsk-kqmlbm]

### PR DESCRIPTION
**Pull request by jaylfc**
CARD TITLE (intent, not commit subject): fold CodeRabbit findings on #2730 (tsk-x6fzgf): Audit lane: add an adversarial-verify stage (refute each finding vs code) before findings become cards

Autonomous build of board card tsk-u2z62g.

REVISION: built on `exec/tsk-x6fzgf` (cut at `14446262fdf55ecbc080d5216e5b144944ad121a`), not on `dev`. That branch's
commits are ancestors of this one. Verified by `git merge-base --is-ancestor`
before the PR was opened.

1. Handle inline and multi-line block comments using lexical state.
2. Exclude hardcoded-secret from generic string-literal filtering.
3. Inspect the actual detector match span for network-exfil instead of the first trigger token.
4. Use a lexical scanner for string-literal classification, handling escapes and backtick literals.

Tests added for each finding; 49 tests pass.

Files:
 changelog.d/tsk-u2z62g-adversarial-verify-fixes.md |   6 +
 changelog.d/tsk-x6fzgf-adversarial-verify.md       |   3 +
 tests/test_code_analyzer.py                        | 110 +++++++++++
 tinyagentos/code_analyzer.py                       | 206 ++++++++++++++++++++-
 4 files changed, 317 insertions(+), 8 deletions(-)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved security analysis accuracy by filtering findings that occur only within comments, string literals, or known example values.
  - Preserved detection of genuine security issues, including hardcoded secrets and network-exfiltration patterns.
  - Improved handling of multiline comments, escaped quotes, template literals, and other complex source-code formatting.

- **Tests**
  - Added comprehensive coverage for false-positive suppression and valid finding detection across common edge cases.

### CODE REVIEW

File: tinyagentos/code_analyzer.py [hunk 1]
```diff
@@ -462,16 +486,182 @@ def analyze_app_source(files: dict[str, str]) -> list[Finding]:
...
+    for f in findings:
+        content = files.get(f.file, "")
+        if not content:
+            verified.append(f)
+            continue
+        lines = content.splitlines()
+        if f.line < 1 or f.line > len(lines):
+            continue
+        line = lines[f.line - 1]
+        block_comment_mask = _compute_block_comment_mask(lines)
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/code_analyzer.py]: Fixed in 2470a6baf. `adversarial_verify` now caches `(lines, mask)` per file in a `lexed` dict and reuses it for every finding in that file, so the lex is once per file instead of once per finding. Regression test `test_block_comment_mask_is_built_once_per_file` monkeypatches `_compute_block_comment_mask` with a counting shim and asserts exactly one build for three findings (three before the fix).

File: tinyagentos/code_analyzer.py [hunk 2]
```diff
@@ -462,16 +486,182 @@ def analyze_app_source(files: dict[str, str]) -> list[Finding]:
...
+    in_block = False
+    mask: list[bool] = []
+    for line in lines:
+        is_in_block = in_block
+        if in_block:
+            if "*/" in line:
+                in_block = False
+        else:
+            if "/*" in line:
+                is_in_block = True
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/code_analyzer.py]: Fixed in 2470a6baf. The mask is now a start-of-line block state produced by a new `_block_state_after_line` scanner that ignores `/*` and `*/` inside string literals and after `//`. The `if in_block_comment: return True` short-circuit in `_trigger_is_inside_string` is gone: the start state is threaded into `_is_inside_string_or_comment`, which carries it forward to the finding's own match position, so code after a same-line `*/` stays eligible. Tests: `test_eval_after_closed_block_comment_on_same_line_is_kept`, `test_eval_before_block_comment_on_same_line_is_kept`, `test_block_comment_open_inside_string_does_not_mask_later_lines`, `test_block_comment_open_after_line_comment_does_not_mask_later_lines` — all red before the change; the existing inside-comment tests stay green.

File: tinyagentos/code_analyzer.py [hunk 3]
```diff
@@ -462,16 +486,182 @@ def analyze_app_source(files: dict[str, str]) -> list[Finding]:
...
+            string_char = ch
+            continue
+
+    return in_string or in_single_comment or in_block
+
+
+def _trigger_is_inside_string(finding: Finding, line: str, in_block_comment: bool = False) -> bool:
+    if in_block_comment:
+        return True
+    if finding.match_start and finding.match_end:
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/code_analyzer.py]: Fixed in 2470a6baf. Replaced the truthiness test with `_has_span(finding)` = `match_end > match_start`, matching the dataclass contract that an unavailable span is `0, 0`. Regression test `test_match_span_starting_at_column_zero_is_used` uses `new Function(x); // eval(y)`: the match starts at offset 0, and before the fix the trigger-token fallback found the `eval(` inside the trailing line comment and refuted the finding.

File: tinyagentos/code_analyzer.py [hunk 4]
```diff
@@ -462,16 +486,182 @@ def analyze_app_source(files: dict[str, str]) -> list[Finding]:
...
+        block_comment_mask is not None and 0 <= line_idx < len(block_comment_mask) and block_comment_mask[line_idx]
+    )
+
+    if finding.rule_id != "hardcoded-secret" and _trigger_is_inside_string(finding, line, in_block_comment=in_block_comment):
+        return True
+
+    if finding.rule_id == "hardcoded-secret":
+        for example in _KNOWN_EXAMPLES:
+            if example in line:
+                return True
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/code_analyzer.py]: Fixed in 2470a6baf. The allowlist now tests `line[match_start:match_end]` rather than the whole line (falling back to the line only when the finding carries no span, which keeps `test_known_example_key_is_refuted` honest). That alone was not sufficient: `detect_hardcoded_secrets` used `pattern.search(line)`, so on a line carrying the example key first the real key was never produced as a finding at all. Switched to `pattern.finditer(line)` so every match on a line is reported. Regression test `test_real_key_beside_example_key_on_same_line_is_kept` asserts the surviving finding's span covers the real key, not the example.

File: tinyagentos/code_analyzer.py [hunk 5]
```diff
@@ -462,16 +486,182 @@ def analyze_app_source(files: dict[str, str]) -> list[Finding]:
...
+            continue
+        line = lines[f.line - 1]
+        block_comment_mask = _compute_block_comment_mask(lines)
+        if _is_clearly_false_positive(f, line, block_comment_mask=block_comment_mask, line_idx=f.line - 1):
+            continue
+        verified.append(f)
+    return verified
+
+
+def _compute_block_comment_mask(lines: list[str]) -> list[bool]:
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/code_analyzer.py:615]: Confirmed and fixed in 2470a6baf, including the `eval(x); /* end */` case you called out. The short-circuit is removed and `_is_inside_string_or_comment` now makes the call from the finding's match span, seeded with a start-of-line block state from the new `_block_state_after_line` scanner (which also stops `/*` inside a string literal or after `//` from masking the rest of the file). The no-span trigger-token fallback additionally scans every occurrence of the token and keeps the finding if any occurrence is live code, instead of judging the line by the first one. Four regression tests cover it, all red beforehand.

File: tinyagentos/code_analyzer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/code_analyzer.py]: Confirmed and fixed in 2470a6baf. `adversarial_verify` now lexes each file once into a `lexed` cache of `(lines, mask)` and reuses it for every finding on that file, so the install/publish path is O(lines + findings). `test_block_comment_mask_is_built_once_per_file` proves it with a counting shim over `_compute_block_comment_mask`.
